File: sys/petscOptionsSetReal.m
function [errCode, toplevel] = petscOptionsSetReal(iname, value)
% Sets an option name-value pair in the options database, overriding
% whatever is already present.
%
%   errCode = petscOptionsSetReal(iname, value) sets the option
%   iname in the option database to the given real value. The iname must start
%   with '-', and must be null-terminated.
%
%   errCode(int) return code (0 indicates OK)
%
% This is a high-level wrapper for petscOptionsSetValue.
%
% SEE ALSO: petscOptionsSetInt, petscOptionsSetValue, petscOptionsGetReal

%#codegen -args {coder.typeof(char(0), [1,inf]), PetscScalar(0)}

errCode = int32(-1);

if ~isempty(coder.target)
    toplevel = nargout>1;
    if ~isempty(iname) && iname(end) && (toplevel || m2c_debug)
        m2c_error('Petsc4m:petscOptionsSetReal:InputError', ...
            'Argument name must be a null-terminated string.')
    end
    
    str = m2c_num2str(value);
    errCode = coder.ceval('PetscOptionsSetValue', PETSC_NULL_OPTIONS, ...
        coder.rref(iname), coder.rref(str));
    
    if errCode && (toplevel || m2c_debug)
        m2c_error('petsc:RuntimeError', 'PetscOptionsSetValue returned error code %d\n', errCode)
    end
end
end

File: sys/codegen/lib/petscOptionsSetReal/petscOptionsSetReal.c
#include "petscOptionsSetReal.h"
#include "petscOptionsSetReal_types.h"
#include "m2c.h"
#include "petsc4m.h"

static void b_m2c_error(int varargin_3);

static void m2c_error(void);

static void b_m2c_error(int varargin_3)
{
  const char *fmt;
  const char *msgid;
  msgid = "petsc:RuntimeError";
  fmt = "PetscOptionsSetValue returned error code %d\n";
  M2C_error(msgid, fmt, varargin_3);
}

static void m2c_error(void)
{
  const char *fmt;
  const char *msgid;
  msgid = "Petsc4m:petscOptionsSetReal:InputError";
  fmt = "Argument name must be a null-terminated string.";
  M2C_error(msgid, fmt);
}

void petscOptionsSetReal(const emxArray_char_T *iname, double value,
                         int *errCode, boolean_T *toplevel)
{
  int i;
  char str[32];
  unsigned char t_str[32];
  const char *iname_data;
  char *ptr;
  iname_data = iname->data;
  *toplevel = true;
  if ((iname->size[1] != 0) && (iname_data[iname->size[1] - 1] != '\x00')) {
    m2c_error();
  }
  ptr = (char *)(&t_str[0]);
  sprintf(ptr, "%.17g", value);
  for (i = 0; i < 32; i++) {
    str[i] = (signed char)t_str[i];
  }
  PetscOptions obj;
  obj = NULL;
  *errCode = PetscOptionsSetValue(obj, &iname_data[0], &str[0]);
  if (*errCode != 0) {
    b_m2c_error(*errCode);
  }
}

void petscOptionsSetReal_initialize(void)
{
}

void petscOptionsSetReal_terminate(void)
{
}
